### ISSUE TITLE
NVDA ignores visually hidden <thead> and/or children when navigating cells

### ISSUE THREAD

**Issue by JamesCatt** [nvaccess/nvda#9879]
### Steps to reproduce:
Author a `<table>` with a `<thead>`, and apply styling to visually hide either the `<thead>` or its child `<tr>` or `<th>` elements. Navigate the `<td>` cells with NVDA + Firefox.

Example table:
```
    <table>
        <thead>
            <tr>
                <th>Column 1</th>
                <th>Column 2</th>
                <th>Column 3</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>Lorem</td>
                <td>ipsum</td>
                <td>dolor</td>
            </tr>
            <tr>
                <td>sit</td>
                <td>amet</td>
                <td>consectetor</td>
            </tr>
        </tbody>
    </table>
```

### Actual behavior:
NVDA does not announce the column headings when navigating the table cells if styling is applied to hide the column heading content visually (i.e., position: absolute, float, display, etc). *Note:* behavior exists regardless of whether styling is applied to the `<thead>`, `<tr>`, or `<th>` elements.

### Expected behavior:
NVDA announces column headings as usual while navigating the table cells, regardless of the CSS (with the exclusion of display: none or visibility: hidden). *Note* this behaves as expected when using `<th>` elements in the `<tbody>`.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2019.1.1

#### Windows version:
Windows 10

#### Name and version of other software in use when reproducing the issue:
Firefox 67.0.4 (64-bit)

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
2018.3.2 - same behavior

**Comment by Adriani90** [comment]
cc: @MarcoZehe

**Comment by MarcoZehe** [comment]
Please provide the exact CSS you use. Without that, there‘s no chance to know if the header cells show up in the tree or not.

**Comment by JamesCatt** [comment]
@MarcoZehe Seems that setting `position: absolute` is all that's required to trigger it. Tried adding various other styles (i.e., `left: -10000px`, `width: 0`, etc) but didn't notice a difference.

Behavior seems to have changed since I opened this issue, however. I just tested again with FF 73.0b6 + NVDA 2019.2.1. Now the column headers are announced when navigating just with up arrow or down arrow, but CTRL + ALT + ARROW doesn't work as expected. Results differ slightly depending on whether the position is applied to `thead`, `tr` or `th`, but generally speaking the column headers cannot be navigated to this way and NVDA announces edge of table when trying to enter a `th`.

Looked at the a11y tree and it does appear that FF is messing up the roles in some places, but not always.

Chrome (79.0.3945.117) has some similar issues, although it looks like the styling is tripping it into calling it a layout table in some instances.

Test page here: https://codepen.io/jcatt/full/QWwZoPB

**Comment by MarcoZehe** [comment]
@jcsteh any ideas?

**Comment by jcsteh** [comment]
For HTML tables (unless the whole table has display: block or similar), Firefox relies on the layout engine to fetch cells given a row and column. Also, in such a table, Firefox also uses the layout engine to determine row and column numbers and headers for any cells which aren't display: block or similar. The problem here is that position: absolute effectively gives the element display: block, which removes table semantics for that element as far as the layout engine is concerned.

Firefox a11y does have code to try to deal with display: block table elements. Unfortunately, this doesn't work well in some cases when you have a table with a mix of both display: block and normal table elements.

I think we somehow need to fall back to the non-layout table calculation code for everything (including the normal elements) if any element composing the table structure is display: block.

**Comment by Adriani90** [comment]
@jcsteh what is the current status on this as far as Firefox is concerned?

@JamesCatt how is this working now with NVDA 2023.3 in Firefox 120 or Chrome and Edge 119? Did you also try to play with the opacity:0 css class to see if it has an impact?

**Comment by jcsteh** [comment]
This should all be fixed in Firefox now. I tested with the test case above and everything seems to work as expected.
